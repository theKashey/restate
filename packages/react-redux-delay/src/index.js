import React, { Component } from 'react';
import reactReduxRestate from 'react-redux-restate';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';
import throttle from 'lodash.throttle';

const makeUpdateFunction = (mode, delay) => {
  let triggerBack = null;
  let touch = null;

  if (mode === 'debounce') {
    touch = debounce(() => triggerBack(), delay);
  } else {
    touch = throttle(() => triggerBack(), delay);
  }
  return trigger => {
    triggerBack = trigger;
    touch();
  };
};

const SECRET = '__delay_secret';
const RenderChildren = ({ children }) => children;

const delay = (delay, mode, options) => WrappedComponent =>
  reactReduxRestate(
    {},
    stores => stores.default,
    (dispatchers, event) => dispatchers.default(event),
    props => ({
      onUpdate: makeUpdateFunction(mode, delay),
      ignoreProps: [SECRET],
      ...options(props),
    }),
  )(WrappedComponent);

// Component model

const ComponentDelay = delay(16, 'debounce', props => ({
  onUpdate: props[SECRET] && makeUpdateFunction(props[SECRET].mode, props[SECRET].delay),
}))(RenderChildren);

const focusSecret = (mode, delay) => ({
  [SECRET]: {
    mode,
    delay,
  },
});

const getSecret = props => focusSecret(props.mode, props.timeout);

class ReduxDelay extends Component {
  constructor(props) {
    super(props);
    this.state = {
      secret: getSecret(props),
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.timeout !== this.props.timeout || nextProps.mode !== this.props.mode) {
      this.setState({ secret: getSecret(nextProps) });
    }
  }

  render() {
    return <ComponentDelay {...this.state.secret}>{this.props.children}</ComponentDelay>;
  }
}

ReduxDelay.propTypes = {
  mode: PropTypes.oneOf(['debounce', 'throttle']),
  timeout: PropTypes.number.isRequired,
  children: PropTypes.node.isRequired,
};

export { ReduxDelay };

export default delay;
