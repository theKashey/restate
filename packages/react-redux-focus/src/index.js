import React from 'react';
import reactReduxRestate from 'react-redux-restate';
import PropTypes from 'prop-types';

const dispatchDefault = (dispatch, event) => dispatch(event);

const SECRET = '__focus_secret';
const RenderChildren = ({children}) => children;

const focus = (composeState, onDispatch = dispatchDefault, options = {}) => WrappedComponent =>
  reactReduxRestate(
    {},
    (stores, props) => composeState(stores.default, props),
    (dispatchers, event, props) => onDispatch(dispatchers.default, event, props),
    () => ({
      ...options,
      ignoreProps: [SECRET, 'componentProps'],
      deeperProps: ['componentProps']
    })
  )(WrappedComponent);

// Component model
const ComponentFocus = focus(
  (store, props) => props[SECRET].composeState(store, props),
  (dispatcher, event, props) => props[SECRET].onDispatch(dispatcher, event, props),
)(RenderChildren);

const focusSecret = {
  [SECRET]: {
    composeState: (store, props) => props.focus(store, props.componentProps),
    dispatch: (dispatcher, event, props) => props.onDispatch(dispatcher, event, props.componentProps),
  },
};

const ReduxFocus = ({focus, onDispatch, children, ...rest}) => (
    <ComponentFocus {...focusSecret} focus={focus} onDispatch={onDispatch} componentProps={rest}>
      {children}
    </ComponentFocus>
  );

ReduxFocus.propTypes = {
  focus: PropTypes.func.isRequired,
  onDispatch: PropTypes.func,
  children: PropTypes.node,
};

export {ReduxFocus};

export default focus;
