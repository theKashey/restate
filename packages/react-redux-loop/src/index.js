import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import loopMiddleware, {ADD_LOOPER, REMOVE_LOOPER} from "./middleware";

class UnconnectedReduxLoop extends PureComponent {
  static propTypes = {
    children: PropTypes.func.isRequired,
    dispatch: PropTypes.func.isRequired,
  };

  componentDidMount() {
    this.props.dispatch({type: ADD_LOOPER, payload: this.loop});
  }

  componentWillUnmount() {
    this.props.dispatch({type: REMOVE_LOOPER, payload: this.loop});
  }

  loop = (event) => {
    this.props.children(event);
  };

  render() {
    return null;
  }
};

const mapDispatch = dispatch => ({
  dispatch
});

const ReduxLoop = connect(null, mapDispatch)(UnconnectedReduxLoop);

const reduxLoop = actionCallback => props => (
  <ReduxLoop>
    {event => actionCallback(event, props)}
  </ReduxLoop>
);

export {
  ReduxLoop,
  loopMiddleware
};

export default reduxLoop;
