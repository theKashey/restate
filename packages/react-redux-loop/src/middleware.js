let loopers = [];

const ADD_LOOPER = 'REACT-REDUX-LOOP/ADD';
const REMOVE_LOOPER = 'REACT-REDUX-LOOP/REMOVE';

const addLooper = looper => loopers.push(looper);
const removeLooper = looper => {loopers = loopers.filter(x => x !== looper)};

const loop = () => next => action => {
  switch (action.type) {
    case ADD_LOOPER:
      return addLooper(action.payload);

    case REMOVE_LOOPER:
      return removeLooper(action.payload);

    default:
      loopers.forEach(looper => looper(action));
      return next(action);
  }
};

export {
  ADD_LOOPER,
  REMOVE_LOOPER
}

export default loop;