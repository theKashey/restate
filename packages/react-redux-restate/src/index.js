import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { createProvider } from 'react-redux';
import reduxRestate from 'redux-restate';
import hoistStatics from 'hoist-react-statics';
import displayName from 'react-display-name';

import reprovide from './reprovide';

const getStores = (stores, props, context) => {
  const result = {};
  Object.keys(stores).forEach(key => {
    const value = stores[key];
    if (typeof value === 'string') {
      const store = props[value] || context[value];
      if (store) {
        result[key] = store;
      } else {
        throw new Error(`restate: unable to found store {$value} for key ${key}`);
      }
    }
  });
  return result;
};

const restate = (baseStores, composeState, routeDispatch, options = {}) => WrappedComponent => {
  const storeKey = options.storeKey || 'store';
  const restateKey = options.restateKey || 'store';

  const Provider = createProvider(restateKey);

  const contextTypes = {
    [storeKey]: PropTypes.any,
  };
  Object.keys(baseStores).forEach(key => {
    const value = baseStores[key];
    if (typeof value === 'string') {
      contextTypes[value] = PropTypes.any;
    }
  });

  class RestateComponent extends Component {
    static contextTypes = contextTypes;

    constructor(props, context) {
      super();
      this.stores = {
        ...getStores(baseStores, this.props, this.context),
        default: props[storeKey] || context[storeKey],
      };
    }

    componentWillMount() {
      this.store = reduxRestate(this.stores, this.composeState, this.routeDispatch, options);
    }

    componentWillUnmount() {
      this.store.unsubscribe();
    }

    composeState = stores => composeState(stores, this.props);
    routeDispatch = (dispatches, event) => routeDispatch(dispatches, event, this.props);

    render() {
      return (
        <Provider store={this.store}>
          <WrappedComponent {...this.props} />
        </Provider>
      );
    }
  }

  hoistStatics(RestateComponent, WrappedComponent);
  RestateComponent.displayName = `reconnected(${displayName(WrappedComponent)})`;

  return RestateComponent;
};

export { reprovide };

export default restate;
