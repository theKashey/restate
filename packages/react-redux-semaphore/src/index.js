import React from 'react';
import PropTypes from 'prop-types';
import reactReduxRestate from 'react-redux-restate';

const SECRET = '__semaphore_secret';
const RenderChildren = ({ children }) => children;

const semaphore = shouldApplyState => WrappedComponent =>
  reactReduxRestate(
    {},
    (stores, props) => (shouldApplyState(stores.default, props) ? stores.default : undefined),
    (dispatchers, event) => dispatchers.default(event),
    () => ({
      ignoreProps: [SECRET]
    })
  )(WrappedComponent);

// Component model
const ComponentSemaphore = semaphore((store, props) => props[SECRET].condition(store, props))(RenderChildren);

const semaphoreSecret = {
  [SECRET]: {
    condition: (stores, props) => props.condition(stores, props.componentProps),
  },
};

const simpleLock = (stores, props) => !props.locked;

const ReduxSemaphore = ({ condition = simpleLock, children, ...rest }) => (
  <ComponentSemaphore {...semaphoreSecret} condition={condition} componentProps={rest}>
    {children}
  </ComponentSemaphore>
);

ReduxSemaphore.propTypes = {
  condition: PropTypes.func,
  children: PropTypes.node,
  locked: PropTypes.bool,
};

export { ReduxSemaphore };

export default semaphore;
