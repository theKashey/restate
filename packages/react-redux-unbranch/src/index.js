import React from 'react';
import reactReduxRestate from 'react-redux-restate';
import shallowequal from 'shallowequal';
import PropTypes from 'prop-types';

const unbranch = () => WrappedComponent =>
  reactReduxRestate(
    {},
    stores => stores.default,
    (dispatchers, event) => dispatchers.default(event),
    props => ({
      onUpdate: trigger => props.update(trigger),
      areStatesEqual: (newState, oldState) => props.compare(newState, oldState, props),
    }),
  )(WrappedComponent);

const RenderChildren = ({ children }) => children;
const Unbranch = unbranch()(RenderChildren);

const update = trigger => trigger();

const clearValues = keys =>
  keys.reduce((acc, key) => {
    acc[key] = undefined;
    return acc;
  }, {});

const compare = (newState, oldState, { ignoreKeys, passKeys, mode = 'ignore' }) => {
  if (mode === 'pass') {
    return passKeys.reduce((acc, key) => acc && newState[key] === oldState[key]);
  }
  return shallowequal({ ...newState, ...clearValues(ignoreKeys) }, { ...oldState, ...clearValues(ignoreKeys) });
};

export const ReduxUnbranch = ({ children, ignoreKeys = [], passKeys = [], mode = 'ignore' }) => (
  <Unbranch update={update} compare={compare} ignoreKeys={ignoreKeys} passKeys={passKeys} mode={mode}>
    {children}
  </Unbranch>
);

ReduxUnbranch.propTypes = {
  ignoreKeys: PropTypes.arrayOf(PropTypes.string),
  passKeys: PropTypes.arrayOf(PropTypes.string),
  mode: PropTypes.string,
  children: PropTypes.node,
};
