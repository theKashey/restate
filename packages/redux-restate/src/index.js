import $$observable from 'symbol-observable';
import shallowequal from 'shallowequal';
import createObservableFor from './observable';
import createSubscription from './subscribe';

const replaceReducer = () => {
  throw new Error('Dont call replaceReducer on synthetic store.');
};

const getDispatches = stores => {
  const result = {
    called: null,
    returnedValue: null,
    next: Object.keys(stores).reduce((acc, storeKey) => {
      acc[storeKey] = event => {
        result.called = storeKey;
        result.returnedValue = stores[storeKey].dispatch(event);
      };
      return acc;
    }, {}),
  };
  return result;
};

const getStates = stores =>
  Object.keys(stores).reduce((acc, storeKey) => {
    acc[storeKey] = stores[storeKey] && stores[storeKey].getState();
    return acc;
  }, {});

const restate = (stores, createState, onDispatch, initialOptions = {}) => {
  let options = initialOptions;
  if (typeof createState !== 'function') {
    throw new Error('Please provide a createState function');
  }

  if (typeof onDispatch !== 'function') {
    throw new Error('Please provide a onDispatch function');
  }

  const nextState = oldState => createState(getStates(stores)) || oldState;

  let currentState = nextState({});

  let updatePending = false;
  const subscriptions = [];
  const { subscribe, trigger } = createSubscription();

  const dispatch = event => {
    const dispatchers = getDispatches(stores);
    onDispatch(dispatchers.next, event);
    if (!dispatchers.called) {
      console.error('No next dispatch was called at', onDispatch);
      throw new Error('No next dispatch called');
    }
    return dispatchers.returnedValue;
  };

  const getState = () => currentState;

  const areStatesEqual = (prev, next) => {
    const areEqual = options.areStatesEqual && options.areStatesEqual(prev, next);
    return typeof areEqual === 'undefined' ? shallowequal(prev, next) : areEqual;
  };

  const onStateUpdate = () => {
    if (options.async) {
      if (!updatePending) {
        updatePending = Promise.resolve().then(() => {
          updatePending = null;
          trigger();
        });
      }
    } else if (options.onUpdate) {
      options.onUpdate(trigger);
    } else {
      trigger();
    }
  };

  const triggerUpdate = () => {
    const lastState = currentState;
    currentState = nextState(currentState);
    if (!areStatesEqual(lastState, currentState)) {
      onStateUpdate();
    }
  };

  const initialize = () => {
    Object.keys(stores)
      .filter(key => stores[key])
      .forEach(storeKey => subscriptions.push(stores[storeKey].subscribe(triggerUpdate)));
  };

  const unsubscribe = () => {
    subscriptions.forEach(unsubscribe => unsubscribe());
  };

  const replaceOptions = newOptions => {
    options = {
      ...options,
      newOptions,
    };
  };

  initialize();

  return {
    dispatch,
    subscribe,
    getState,
    replaceReducer,
    [$$observable]: createObservableFor(subscribe, getState),

    unsubscribe,
    update: triggerUpdate,
    replaceOptions,
  };
};

export default restate;
