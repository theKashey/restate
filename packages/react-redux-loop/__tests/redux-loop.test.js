import React, {Component} from 'react';
import Enzyme, {mount} from 'enzyme';
import {Provider, connect} from 'react-redux';
import Adapter from 'enzyme-adapter-react-16';
import {createStore, combineReducers, applyMiddleware, compose} from 'redux';
import {ReduxLoop, loopMiddleware} from '../src';

Enzyme.configure({adapter: new Adapter()});


describe('React-redux-loop', () => {

  const list = (state, action) => {
    if (typeof state === 'undefined') {
      return [];
    }

    if (action.type === 'ACTION') {
      return [...state, action.payload];
    }
    return state
  };

  const reducer = combineReducers({
    list
  });

  const store = createStore(
    reducer,
    compose(
      applyMiddleware(loopMiddleware),
    )
  );

  it('should loop events', () => {

    class LoopTest extends Component {

      state = {
        list: []
      };

      componentDidMount() {
        this.props.emit();
      }

      render() {
        return (
          <div>
            Redux list:
            <div className="output">
              <ul>
                {this.props.list.map(x => <li>{x}</li>)}
              </ul>
              React reflection:
              <ul>
                {this.state.list.map(x => <li>{x}</li>)}
              </ul>
            </div>
            <ReduxLoop>
              {action => {
                this.setState({list: [...this.state.list, `${  action.type  }#${  action.payload}`]}, () => {
                  if (this.state.list.length < 10) {
                    this.props.emit(this.state.list.length);
                  }
                });
              }}
            </ReduxLoop>
          </div>
        );
      }
    }

    const mapStateToProps = state => ({list: state.list});
    const mapDispatchToProps = {
      emit: (id) => ({type: 'ACTION', payload: id})
    };

    const ConnectedLoopTest = connect(mapStateToProps, mapDispatchToProps)(LoopTest);

    const wrapper = mount(<Provider store={store}><ConnectedLoopTest/></Provider>);

    console.log(wrapper.debug());

    expect(wrapper.find(LoopTest).prop('list').length).toBe(10);
    expect(store.getState()).toMatchSnapshot();
  });
});
