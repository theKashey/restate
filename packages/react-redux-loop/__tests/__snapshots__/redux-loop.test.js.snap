// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React-redux-loop should loop events 1`] = `
Object {
  "list": Array [
    undefined,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
  ],
}
`;
